fdbserver/masterserver.actor.cpp

```diff
@@ -147,11 +154,46 @@ ACTOR Future<Void> getVersion(Reference<MasterData> self, GetCommitVersionReques
 			if (BUGGIFY) {
 				t1 = self->lastVersionTime;
 			}
-			rep.prevVersion = self->version;
-			self->version +=
+
+			// Versions should roughly follow wall-clock time, based on the
+			// system clock of the current machine and an FDB-specific epoch.
+			// Calculate the expected version and determine whether we need to
+			// hand out versions faster or slower to stay in sync with the
+			// clock.
+			Version toAdd =
 			    std::max<Version>(1,
 			                      std::min<Version>(SERVER_KNOBS->MAX_READ_TRANSACTION_LIFE_VERSIONS,
 			                                        SERVER_KNOBS->VERSIONS_PER_SECOND * (t1 - self->lastVersionTime)));
+			Version expectedVersion = g_network->timer() * SERVER_KNOBS->VERSIONS_PER_SECOND - self->referenceVersion;
+
+			// How far the new version is off from the expected version.
+			Version offset = self->version + toAdd - expectedVersion;
+			// Clamp the offset to a specific range. This prevents huge version
+			// jumps.
+			Version clamped =
+			    std::clamp(offset, -SERVER_KNOBS->MAX_VERSION_RATE_OFFSET, SERVER_KNOBS->MAX_VERSION_RATE_OFFSET);
+
+			// Translate the version offset (in the range [-1,000,000,
+			// 1,000,000]) to a value in the range [0.9, 1.1]). Note that these
+			// values are examples, and are based on the
+			// MAX_VERSION_RATE_OFFSET and MAX_VERSION_RATE_MODIFIER knobs.
+			double modifier =
+			    ((1.0 + SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER) - (1.0 - SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER)) /
```

Comment:
Does `modifier` reduce to `SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER/SERVER_KNOBS->MAX_VERSION_RATE_OFFSET`?

Reply:
Thanks, yes it looks like it does!